server-spi-private/src/main/java/org/keycloak/validation/ValidationContext.java

```diff
@@ -0,0 +1,203 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.validation;
+
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Denotes a context in which the Validation takes place. A {@link ValidationContext} should be created for a
+ * batch of validation checks for a given entity.
+ */
+public class ValidationContext {
+
+    // user registration, user profile update, client registration, realm creation
+    private final ValidationContextKey contextKey;
+
+    // additional context specific attributes
+    private final Map<String, Object> attributes;
+
+    /**
+     * Holds the current {@link RealmModel}
+     */
+    private final RealmModel realm;
+
+    /**
+     * Holds the current {@link UserModel}
+     */
+    private final UserModel user;
+
+    /**
+     * Holds the current {@link ClientModel}
+     */
+    private final ClientModel client;
+
+    /**
+     * If {@literal true} stop on first {@link ValidationProblem.Severity#ERROR}, otherwise perform all validations.
+     */
+    private final boolean failFast;
+
+    /**
+     * Exposes the {@link ValidationActionKey} e.g. default, create, update etc.
+     */
+    private final ValidationActionKey actionKey;
+
+    // TODO add support to skip certain validations for ValidationKey
+
+    public ValidationContext(ValidationContext that) {
+        this(that.getContextKey(), that.getAttributes(), that.getRealm(), that.getUser(), that.getClient(), that.isFailFast(), that.getActionKey());
+    }
+
+    public ValidationContext(ValidationContextKey contextKey) {
+        this(contextKey, Collections.emptyMap(), null, null, null, false, ValidationActionKey.DEFAULT);
+    }
+
+    public ValidationContext(ValidationContextKey contextKey, RealmModel realm) {
+        this(contextKey, Collections.emptyMap(), realm, null, null, false, ValidationActionKey.DEFAULT);
+    }
+
+    public ValidationContext(ValidationContextKey contextKey, Map<String, Object> attributes) {
+        this(contextKey, attributes, null, null, null, false, ValidationActionKey.DEFAULT);
+    }
+
+    /**
+     * @param realm
+     * @param contextKey
+     * @param attributes
+     * @param user
+     * @param client
+     * @param failFast
+     */
+    public ValidationContext(ValidationContextKey contextKey, Map<String, Object> attributes, RealmModel realm, UserModel user, ClientModel client, boolean failFast, ValidationActionKey actionKey) {
```

Comment:
We could add a ctor variant that accepts a KeycloakSession, with this we could access Realm and Client information via `session.getContext()`